src/multiplex/multiplex.rs

```diff
@@ -377,75 +451,175 @@ impl<S, T, E> Multiplex<S, T>
-        match message {
-            Ok(Message::WithoutBody(val)) => {
-                trace!("got in_flight value without body");
-                try!(self.transport.write(Frame::Message(id, val)));
+    fn write_in_message(&mut self,
+                        id: RequestId,
+                        message: Message<T::In, T::Stream>)
+                        -> io::Result<()>
+    {
+        let (message, body) = match message {
+            Message::WithBody(message, rx) => (message, Some(rx)),
+            Message::WithoutBody(message) => (message, None),
+        };
+
+        // Create the frame
+        let frame = Frame::Message {
+            id: id,
+            message: message,
+            body: body.is_some()
+        };
+
+        // Write the frame
+        try!(self.dispatch.transport().write(frame));
+        self.blocked_on_flush.wrote_frame();
+
+        match self.exchanges.entry(id) {
+            Entry::Occupied(mut e) => {
+                assert!(!e.get().responded, "invalid exchange state");
+                assert!(e.get().is_outbound());
+
+                // Track that the exchange has been responded to
+                e.get_mut().responded = true;
+
+                // Set the body receiver
+                e.get_mut().in_body = body;
+
+                // If the exchange is complete, clean up the resources
+                if e.get().is_complete() {
+                    e.remove();
+                }
+            }
+            Entry::Vacant(e) => {
+                // Create the exchange state
+                let exchange = Exchange::new(
+                    Request::In,
+                    self.frame_buf.deque());
+
+                // Track the exchange
+                e.insert(exchange);
             }
-            Ok(Message::WithBody(val, body)) => {
-                trace!("got in_flight value with body");
+        }
 
-                // Write the in frame
-                try!(self.transport.write(Frame::Message(id, val)));
+        Ok(())
+    }
 
-                // Store the body stream for processing
-                self.in_bodies.insert(id, body);
-            }
-            Err(e) => {
-                trace!("got in_flight error");
-                try!(self.transport.write(Frame::Error(id, e)));
-            }
+    fn write_in_error(&mut self,
+                      id: RequestId,
+                      error: T::Error)
+                      -> io::Result<()>
+    {
+        if let Entry::Occupied(mut e) = self.exchanges.entry(id) {
+            assert!(!e.get().responded, "exchange already responded");
+
+            // TODO: should the outbound body be canceled? In theory, if the
+            // consuming end doesn't want it anymore, it should drop interest
+            e.get_mut().responded = true;
+            e.get_mut().out_body = None;
+            e.get_mut().in_body = None;
+            e.get_mut().out_deque.clear();
```

Comment:
This is all covered by `e.remove()` below, right? Does this need to explicitly happen still?